### DISCUSSION TITLE
TypeAdapter validation of model with generics accepts only dictionary

### DISCUSSION THREAD

**Discussion by ohad-weka** [pydantic/pydantic#6090] [Pydantic V2]
There seem to be a difference between validation model without generics vs one with -

```py
from typing import Generic, TypeVar

from pydantic import BaseModel

T = TypeVar('T')

class M(BaseModel, Generic[T]):
    foo: T

# Ok
TypeAdapter(M[list]).validate_python(dict(foo=[1, 2, 3]))
```
```py
# Error
TypeAdapter(M[list]).validate_python(M(foo=[1, 2, 3]))
```
```yaml
pydantic_core._pydantic_core.ValidationError: 1 validation error for M[list]
  Input should be a valid dictionary [type=dict_type, input_value=M(foo=[1, 2, 3]), input_type=M]
    For further information visit https://errors.pydantic.dev/0.30.0/v/dict_type
```

For a model without generics both ways supplying a dict/model object work fine - is this the intended behavior/bug/other?

**Comment by adriangb** [comment]
Seems to work fine on [3511703](https://github.com/pydantic/pydantic/commit/3511703fc2d2f5769732aef53a190c344856794a):

```python
from typing import Generic, TypeVar

from pydantic import BaseModel, TypeAdapter

T = TypeVar('T')

class M(BaseModel, Generic[T]):
    foo: T


ta = TypeAdapter(M[list[int]])

v = ta.validate_python({'foo': [1, 2, 3]})
print(v)
#> foo=[1, 2, 3]
```

**Reply by ohad-weka** [reply]
@adriangb Thanks for the reply! 
I see now the requirement, this does work - 

```py
TypeAdapter(M[list]).validate_python(M[list](foo=[1, 2]))
```
The _list_ generic must be mentioned where this fails - 
```py
TypeAdapter(M[list]).validate_python(M(foo=[1, 2]))
```
Although it is actually just "like" the first one I'd say.
Is this deliberate or makes sense to change?

**Reply by adriangb** [reply]
@dmontagu any thoughts?

**Reply by ohad-weka** [reply]
Bumping

**Reply by dmontagu** [reply]
Yeah, we talked about making it so that if models have the same generic origin that they are allowed for validation by pydantic core. Or at least dump it to a dict in that case prior to parsing. That would resolve this issue, and I think is preferable in general, but I'm not sure how much work it would be in pydantic-core.

In the meantime, you can do:
```python
from typing import Generic, TypeVar, Annotated

from pydantic import BaseModel, TypeAdapter, WrapValidator

T = TypeVar('T')


class M(BaseModel, Generic[T]):
    foo: T


def dump_matching_generic(v, h):
    if isinstance(v, M):
        v = v.model_dump()
    return h(v)


ta = TypeAdapter(Annotated[M[list[int]], WrapValidator(dump_matching_generic)])

v = ta.validate_python(M(foo=[1, 2]))
print(v)
# > foo=[1, 2]
```

If you want to speed up integration of a better solution into the library, I would suggest creating an issue (not a discussion) as a feature request. PR welcome to make the change but I think it may require digging into the pydantic-core internals a bit.

**Reply by ohad-weka** [reply]
Thanks @dmontagu I'll open an issue
